File: 290/Solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <sstream>

using namespace std;


class Solution {

public:
    bool wordPattern(const string &pattern , const string &str){
        vector<string> words;
        stringstream ss(str);
        string word;

        while (ss >> word){
            words.push_back(word);
        }

        if (words.size() != pattern.size()) return false;

        unordered_map<char, int> charToIndex;
        unordered_map<string, int> stringToIndex;

        for (int i = 0; i < pattern.size(); ++i) {
            char c = pattern[i];
            string w = words[i];
            
            // 检查字符和单词的映射是否一致
            if (charToIndex.find(c) == charToIndex.end()) {
                charToIndex[c] = i;
            }
            if (wordToIndex.find(w) == wordToIndex.end()) {
                wordToIndex[w] = i;
            }
            
            if (charToIndex[c] != wordToIndex[w]) {
                return false;
            }
        }
        
        return true;

    }

};

File: 290/Solution.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    public boolean wordPattern(String pattern, String str) {
        String[] strs = str.split(" ");
        if (pattern.length() != str.length()) return false;
        Map<Object,Integer> map = new HashMap<>();
        for (Integer i = 0 ; i < pattern.length() ; i++){
            if (map.put(pattern.charAt(i),i) != map.put(strs[i],i)){
                return false;
            }
        }
        return true;
    }
}